### PR TITLE
[DYOD] Allow for multi-column sort

### PR DESCRIPTION
**Pull request by mrks**
This supersedes #2005, which I somehow broke when I pushed my refactorings to a remote repository. I addressed hopefully all review comments from there. Since some of these changes touched more than a single line, I'd appreciate a review from someone who was not involved in this.

```diff
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | Benchmark      | prev. iter/s       | runs  | new iter/s         | runs  | change [%] | p-value |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
 | TPC-H 01       | 1.2757476568222046 | 77    | 1.2572546005249023 | 76    | -1%        |  0.0000 |
 | TPC-H 02       | 79.53157043457031  | 4772  | 80.58978271484375  | 4836  | +1%        |  0.0889 |
 | TPC-H 03       | 8.908345222473145  | 535   | 9.315204620361328  | 559   | +5%        |  0.0000 |
+| TPC-H 04       | 7.828646183013916  | 470   | 8.386175155639648  | 504   | +7%        |  0.0000 |
+| TPC-H 05       | 3.8665037155151367 | 232   | 4.401660442352295  | 265   | +14%       |  0.0000 |
-| TPC-H 06       | 264.18682861328125 | 15852 | 247.61444091796875 | 14857 | -6%        |  0.0000 |
 | TPC-H 07       | 9.634734153747559  | 579   | 9.964807510375977  | 598   | +3%        |  0.0000 |
+| TPC-H 08       | 7.403852462768555  | 445   | 7.813937664031982  | 469   | +6%        |  0.0000 |
 | TPC-H 09       | 1.7570385932922363 | 106   | 1.810500979423523  | 109   | +3%        |  0.0000 |
 | TPC-H 10       | 3.362816333770752  | 202   | 3.4487195014953613 | 207   | +3%        |  0.0088 |
+| TPC-H 11       | 38.01150131225586  | 2281  | 41.288909912109375 | 2478  | +9%        |  0.0000 |
 | TPC-H 12       | 14.9895658493042   | 900   | 15.060484886169434 | 904   | +0%        |  0.0001 |
+| TPC-H 13       | 2.2825334072113037 | 137   | 2.6715123653411865 | 161   | +17%       |  0.0000 |
 | TPC-H 14       | 44.687896728515625 | 2682  | 43.30529022216797  | 2599  | -3%        |  0.0000 |
-| TPC-H 15       | 118.19493103027344 | 7092  | 111.2449951171875  | 6675  | -6%        |  0.0000 |
+| TPC-H 16       | 9.052909851074219  | 544   | 10.17762565612793  | 611   | +12%       |  0.0000 |
+| TPC-H 17       | 5.363152027130127  | 322   | 5.804923057556152  | 349   | +8%        |  0.0000 |
 | TPC-H 18       | 1.2516614198684692 | 76    | 1.2373242378234863 | 75    | -1%        |  0.0324 |
+| TPC-H 19       | 9.902627944946289  | 595   | 10.5778169631958   | 635   | +7%        |  0.0000 |
+| TPC-H 20       | 31.53620147705078  | 1893  | 33.549686431884766 | 2013  | +6%        |  0.0000 |
 | TPC-H 21       | 1.3884423971176147 | 84    | 1.4122768640518188 | 85    | +2%        |  0.0002 |
 | TPC-H 22       | 18.706771850585938 | 1123  | 17.926311492919922 | 1076  | -4%        |  0.0000 |
 | geometric mean |                    |       |                    |       | +4%        |         |
 +----------------+--------------------+-------+--------------------+-------+------------+---------+
```

There are some additional optimizations possible if we return ReferenceSegments instead of ValueSegments where possible. Those will not be part of this PR.

Screenshot 2020-03-29 at 15 32 08

### CODE REVIEW

File: src/lib/operators/sort.cpp [hunk 1]
```diff
@@ -1,260 +1,283 @@
 #include "sort.hpp"
```
Comment (mrks) [inline src/lib/operators/sort.cpp]: The Github diff here is a complete mess - I'd just look at the new file.

File: src/lib/operators/sort.cpp [hunk 2]
```diff
@@ -1,260 +1,283 @@
...
 
-        for (auto row_index = 0u; row_index < row_count_out; ++row_index) {
-          const auto [chunk_id, chunk_offset] = _row_id_value_vector->at(row_index).first;
+namespace opossum {
 
-          auto& segment_ptr_and_typed_ptr_pair = segment_ptr_and_accessor_by_chunk_id[chunk_id];
-          auto& base_segment = segment_ptr_and_typed_ptr_pair.first;
-          auto& accessor = segment_ptr_and_typed_ptr_pair.second;
+Sort::Sort(const std::shared_ptr<const AbstractOperator>& in, const std::vector<SortColumnDefinition>& sort_definitions,
+           const size_t output_chunk_size)
```
Comment (Bensk1) [inline src/lib/operators/sort.cpp]: see above

File: src/lib/operators/sort.cpp [hunk 3]
```diff
@@ -1,260 +1,283 @@
...
+  for (const auto& column_sort_definition : _sort_definitions) {
+    Assert(column_sort_definition.column != INVALID_COLUMN_ID, "Sort: Invalid column in sort definition");
+    Assert(column_sort_definition.column < input_table->column_count(),
+           "Sort: Column ID is greater than table's column count");
+  }
 
-    return output;
+  std::shared_ptr<Table> sorted_table;
+
+  // Starting after the first (least significant) sort operation, this holds the order of the table as it has been
```
Comment (Bensk1) [inline src/lib/operators/sort.cpp]: "After the first (least significant) sort operation has been completed..."?

File: src/lib/operators/sort.cpp [hunk 4]
```diff
@@ -1,260 +1,283 @@
...
+  // determined so far. This is not a completely proper PosList on the input table as it might point to
+  // ReferenceSegments.
+  auto previously_sorted_pos_list = std::optional<PosList>{};
+
+  for (auto sort_step = static_cast<int64_t>(_sort_definitions.size() - 1); sort_step >= 0; --sort_step) {
+    const bool is_last_sorting_run = (sort_step == 0);
```
Comment (Bensk1) [inline src/lib/operators/sort.cpp]: `is_last_sorting_step`?

File: src/lib/operators/sort.cpp [hunk 5]
```diff
@@ -1,260 +1,283 @@
...
+    });
   }
 
- protected:
-  const std::shared_ptr<const Table> _table_in;
-  const size_t _output_chunk_size;
-  const std::shared_ptr<std::vector<std::pair<RowID, SortColumnType>>> _row_id_value_vector;
-};
+  auto final_sort_definition = _sort_definitions[0];
+  // Set the ordered_by attribute of the output's chunks to the column the chunk was sorted by last (because this
```
Comment (Bensk1) [inline src/lib/operators/sort.cpp]: I could imagine removing the whole part in braces and replacing it somehow with above's "most significant".

File: src/lib/operators/sort.cpp [hunk 6]
```diff
@@ -1,260 +1,283 @@
...
   const ColumnID _column_id;
   const OrderByMode _order_by_mode;
-  // chunk size of the materialized output
-  const size_t _output_chunk_size;
 
-  std::shared_ptr<std::vector<RowIDValuePair>> _row_id_value_vector;
-  std::shared_ptr<std::vector<RowIDValuePair>> _null_value_rows;
+  std::vector<RowIDValuePair> _row_id_value_vector;
+
+  // Stored as RowIDValuePair even if value is unused for better type compatibility.
```
Comment (Bensk1) [inline src/lib/operators/sort.cpp]: Shouldn't that be "Stored as RowIDValuePair for better type compatibility even if value is unused."?

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (mrks) [reply src/lib/operators/sort.cpp]: > Set the ordered_by attribute of the output's chunks according to the most significant sort operation, which is the column the table was sorted by last.

should be enough, don't you thnk?

**Comment by mrks** [comment]
Good points.

File: src/lib/operators/sort.cpp (hunk 5 above)
Comment (Bensk1) [reply src/lib/operators/sort.cpp]: Yes.

**Pull request merged**